File: src/main/scala/codacy/rubocop/Generator.scala
package codacy.rubocop

import java.io.File

import com.fasterxml.jackson.core.JsonParser.Feature
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.dataformat.yaml.YAMLFactory
import com.typesafe.config.ConfigFactory
import play.api._
import play.api.libs.json._
import play.api.libs.ws._
import play.api.libs.ws.ning._

import scala.concurrent.Await
import scala.concurrent.duration._

object Generator {

  def main(args: Array[String]): Unit = {
    val version = args.headOption.getOrElse {
      throw new Exception("Version is required")
    }

    val filesName = Set("default", "disabled", "enabled")
    val mergedJson = withClient { implicit client =>
      filesName.map { fileName =>
        val cops = getConfig(version, fileName)

        def yamlReader: ObjectMapper = new ObjectMapper(new YAMLFactory())
          .configure(Feature.ALLOW_COMMENTS, true)
          .configure(Feature.AUTO_CLOSE_SOURCE, true)
          .configure(Feature.ALLOW_UNQUOTED_FIELD_NAMES, true)
          .configure(Feature.ALLOW_SINGLE_QUOTES, true)
          .configure(Feature.ALLOW_UNQUOTED_CONTROL_CHARS, true)

        Json.parse(yamlReader.readTree(cops).toString).as[JsObject]
      }.foldLeft(Json.obj()) { case (jsonAcc, json) => jsonAcc.deepMerge(json) }
    }
    val patterns = generatePatterns(mergedJson)
    val patternsJsonString = Json.prettyPrint(
      Json.obj(
        "name" -> "Rubocop",
        "version" -> version,
        "patterns" -> Json.parse(Json.toJson(patterns).toString).as[JsArray]
      )
    )

    val oldParamDescriptions: Map[String, JsArray] = ResourceHelper.getResourceContent("docs/description/description.json")
      .toOption.fold(Map.empty[String, JsArray]) { lines =>
      Json.parse(lines.mkString("\n"))
        .as[Array[JsObject]]
        .flatMap { obj =>
          obj.\("parameters").asOpt[JsArray].fold(Option.empty[(String, JsArray)]) { parameters =>
            Option((obj.\("patternId").as[String], parameters))
          }
        }(collection.breakOut)
    }
    val descriptions = generateDescriptions(mergedJson, oldParamDescriptions)
    val descriptionsJsonString = Json.prettyPrint(Json.parse(Json.toJson(descriptions).toString).as[JsArray])

    val repoRoot = new java.io.File(".")
    val docsRoot = new java.io.File(repoRoot, "src/main/resources/docs")
    val patternsFile = new java.io.File(docsRoot, "patterns.json")
    val descriptionsRoot = new java.io.File(docsRoot, "description")
    val descriptionsFile = new java.io.File(descriptionsRoot, "description.json")

    ResourceHelper.writeFile(patternsFile.toPath, patternsJsonString)
    ResourceHelper.writeFile(descriptionsFile.toPath, descriptionsJsonString)
  }

  private def withClient[T](f: NingWSClient => T): T = {
    val configuration = Configuration.reference ++
      Configuration(ConfigFactory.parseString(
        """
          |ws.followRedirects = true
        """.stripMargin)
      )
    val environment = Environment(new File("."), this.getClass.getClassLoader, Mode.Prod)
    val parser = new WSConfigParser(configuration, environment)
    val config = NingWSClientConfig(wsClientConfig = parser.parse())
    val asyncConfig = new NingAsyncHttpClientConfigBuilder(config).build()
    val client = NingWSClient(asyncConfig)

    val res = f(client)

    client.close()

    res
  }

  private def getConfig(version: String, fname: String)(implicit client: NingWSClient): String = {
    val url = s"https://raw.githubusercontent.com/bbatsov/rubocop/v$version/config/$fname.yml"
    Await.result(client.url(url).get(), 30.seconds).body
  }

  private def generatePatterns(json: JsObject): JsArray = {
    val allPatterns = json.value.collect {
      case (pattern, JsObject(obj)) =>
        val specs = obj.filterNot {
          case (key, _) =>
            key.startsWith("Supported") ||
              key.startsWith("StyleGuide") ||
              key.startsWith("Enabled") ||
              key.startsWith("Description") ||
              key.startsWith("Reference") ||
              key.startsWith("AutoCorrect")
        }

        (pattern.replace("/", "_"), specs)
    }

    val codacyPatterns = allPatterns.collect {
      case (pattern, parameterMap) =>
        val level = if (pattern.startsWith("Lint") ||
          pattern.startsWith("Metrics") ||
          pattern.startsWith("Performance") ||
          pattern.startsWith("Rails")) {
          "Warning"
        } else {
          "Info"
        }

        val category = if (pattern.startsWith("Lint") ||
          (pattern.startsWith("Metrics") && !pattern.contains("Length") && !pattern.contains("AbcSize")) ||
          pattern.startsWith("Rails")) {
          "ErrorProne"
        } else if (pattern.startsWith("Performance")) {
          "Performance"
        } else if (pattern.startsWith("Security")) {
          "Security"
        } else {
          "CodeStyle"
        }

        val basePattern = Json.obj(
          "patternId" -> pattern,
          "level" -> level,
          "level" -> level,
          "category" -> category
        )

        if (parameterMap.keys.nonEmpty) {
          val params: List[JsObject] = parameterMap.map { case (key, value) => Json.obj("name" -> key, "default" -> value) }(collection.breakOut)
          val paramsMap: JsArray = Json.arr(params.map(Json.toJsFieldJsValueWrapper(_)): _*)

          basePattern ++ Json.obj(
            "parameters" -> paramsMap
          )
        } else {
          basePattern
        }
    }

    Json.parse(Json.toJson(codacyPatterns).toString).as[JsArray]
  }

  private def generateDescriptions(json: JsObject, oldParamDescriptions: Map[String, JsArray]): JsArray = {
    val allPatternsWithDescs = json.value.collect {
      case (pattern, JsObject(obj)) =>
        val descriptions = obj.filterNot {
          case (key, _) =>
            key.startsWith("Supported") ||
              key.startsWith("StyleGuide") ||
              key.startsWith("Enabled") ||
              key.startsWith("Reference") ||
              key.startsWith("AutoCorrect")
        }

        (pattern.replace("/", "_"), descriptions)
    }

    val codacyPatternsDescs = allPatternsWithDescs.collect {
      case (pattern, parameterMap) if parameterMap.get("Description").isDefined =>
        val params = oldParamDescriptions.get(pattern).fold(Json.obj())(params => Json.obj(("parameters", params)))

        val basePattern = Json.obj(
          "patternId" -> pattern,
          "title" -> parameterMap("Description"),
          "description" -> parameterMap("Description"),
          "timeToFix" -> 5
        ) ++ params
        basePattern
    }

    Json.parse(Json.toJson(codacyPatternsDescs).toString).as[JsArray]
  }

}

File: src/main/ruby/codacy/rubocop/generator.rb
#!/usr/bin/env ruby

# frozen_string_literal: true

require 'yaml'
require 'json'

module RubocopDoc
  module Codacy
    module Markdown
      def self.generate_markdown_file(file_name, content)
        File.write("src/main/resources/docs/description/#{file_name}.md", content)
      end

      def self.file_name(cop_data)
        cop_data[:name].gsub("/", "_")
      end

      def self.content(cop_data)
        "\n#{cop_data[:description]}\n\n# Examples\n\n#{cop_data[:examples_description]}\n[Source](#{info_url(cop_data)})"
      end

      def self.info_url(cop_data)
        "http://www.rubydoc.info/gems/rubocop/RuboCop/Cop/#{cop_data[:name]}"
      end

      def self.run(file_path = "rubocop-doc.yml")
        cops_data = YAML.load_file(file_path)
        cops_data.each do |cop_data|
          generate_markdown_file(
            file_name(cop_data),
            content(cop_data)
          )
        end
      end
    end

    module DescriptionJSON
      def self.generate_json_file(hash)
        File.write("src/main/resources/docs/description/description.json", JSON.pretty_generate(hash))
      end

      def self.run(file_path = "rubocop-doc.yml")
        cops_data    = YAML.load_file(file_path)
        descriptions = cops_data.map do |cop_data|
          {
            patternId:   cop_data[:name].gsub("/", "_"),
            title:       cop_data[:configuration]['Description'],
            description: cop_data[:configuration]['Description'],
            timeToFix:   5
          }
        end
        generate_json_file(descriptions)
      end

    end

    module PattersJSON
      def self.generate_json_file(hash)
        File.write("src/main/resources/docs/patterns.json", JSON.pretty_generate(hash))
      end

      def self.level(cop_data)
        if ["Security"].include?(cop_data[:department_name])
          "Error"
        elsif ["Lint", "Performance", "Rails"].include?(cop_data[:department_name])
          "Warning"
        elsif cop_data[:department_name].start_with?("Metrics") &&
          !cop_data[:department_name].include?("Length") &&
          !cop_data[:department_name].include?("AbcSize")
          "Warning"
        else
          "Info"
        end
      end

      def self.category(cop_data)
        if ["Lint"].include?(cop_data[:department_name])
          "ErrorProne"
        elsif cop_data[:department_name].start_with?("Metrics") &&
          !cop_data[:department_name].include?("Length") &&
          !cop_data[:department_name].include?("AbcSize")
          "ErrorProne"
        elsif ["Rails"].include?(cop_data[:department_name])
          "ErrorProne"
        elsif ["Performance"].include?(cop_data[:department_name])
          "Performance"
        elsif ["Security"].include?(cop_data[:department_name])
          "Security"
        else
          "CodeStyle"
        end
      end

      def self.subcategory(cop_data)
        if category(cop_data) == "Security"
          case cop_data[:name] 
          when "Security/JSONLoad", "Security/MarshalLoad"
            "InsecureModulesLibraries"
          when "Security/Open"
            "CommandInjection"
          when "Security/YAMLLoad", "Security/Eval"
            "InputValidation"
          else nil
          end
        end
      end

      def self.parameters(cop_data)
        cop_data[:configurable_attributes].map do |key, value|
          { name: key, default: value }
        end
      end

      def self.run(file_path = "rubocop-doc.yml")
        cops_data = YAML.load_file(file_path)
        patterns  = cops_data.map do |cop_data|
          {
            patternId:  cop_data[:name].gsub("/", "_"),
            level:      level(cop_data),
            category:   category(cop_data),
            subcategory: subcategory(cop_data),
            parameters: parameters(cop_data)
          }.delete_if { |key, value| value == nil }
        end
        data      = {
          name:     "Rubocop",
          version:  File.read(".rubocop-version").strip,
          patterns: patterns
        }
        generate_json_file(data)
      end
    end
  end
end

RubocopDoc::Codacy::Markdown.run
RubocopDoc::Codacy::PattersJSON.run
RubocopDoc::Codacy::DescriptionJSON.run
